### DISCUSSION TITLE
[WASM] Profile guided AOT - Failed to clone the Emscripten SDK repository.

### DISCUSSION THREAD

**Discussion by asadullah-refat-89** [unoplatform/uno#11790] [Q&A] [answered]
### Current behavior

I followed the official documentation to create an AOT profile to enable profile-guided AOT.

In the Wasm csproj, I added the following:
`<WasmShellGenerateAOTProfile>true</WasmShellGenerateAOTProfile>`

In the LinkerConfig.xml file, I added the following:
`<assembly fullname="WebAssembly.Bindings" />`

I pressed `Ctrl+F5` to start building.

Then I get the following error:

`C:\Users\User\.nuget\packages\uno.wasm.bootstrap\7.0.19\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.1270310] Failed to clone the Emscripten SDK repository. Review the error logs for additional information and check https://aka.platform.uno/msbuild-troubleshoot`

`C:\Users\User\.nuget\packages\uno.wasm.bootstrap\7.0.19\build\Uno.Wasm.Bootstrap.targets(207,3): error : System.Exception: Failed to run emscripten. More details are available in diagnostics logging mode or using the MSBuild /bl switch, see https://aka.platform.uno/msbuild-troubleshoot.`

`C:\Users\User\.nuget\packages\uno.wasm.bootstrap\7.0.19\build\Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_vfa2b0f7b8718563b261ab09643f013eb7a68a3ba.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 725`

`C:\Users\User\.nuget\packages\uno.wasm.bootstrap\7.0.19\build\Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_vfa2b0f7b8718563b261ab09643f013eb7a68a3ba.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 257`

### Expected behavior

The expected behavior would be that the build succeeds and I can save the AOT profile by pressing `Alt+Shift+P`.

### How to reproduce it (as minimally and precisely as possible)

Clone the repo and follow the steps to generate an AOT profile and you will see the errors mentioned.
Link to the repo: https://github.com/asadullahrifat89/honk-trooper-uno-platform

### Workaround

NA

### Works on UWP/WinUI

NA

### Environment

Release

### NuGet package version(s)

Uno.Wasm.Bootstrap 7.0.19

### Affected platforms

WebAssembly

### IDE

Visual Studio 2022

### IDE version

17.5.3

### Anything else we need to know?

Help would be much appreciated.

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report. Give a look at the link provided by the error, you will likely see additional information.

**Reply by asadullah-refat-89** [reply]
I have turned on detailed build verbosity and found this:

`[00:00:17.3499899] [229/355] [EMCC] corebindings.c -> corebindings.o
2>    [00:00:17.3500170] FAILED: corebindings.o
2>    [00:00:17.3500206] powershell "$env:PYTHONUTF8=1; emcc  -DENABLE_METADATA_UPDATE=1 -s USE_ZLIB=1 -s FORCE_FILESYSTEM=1  -fwasm-exceptions -DGEN_PINVOKE -IC:/Users/User/AppData/Local/Temp/dotnet-runtime-wasm-windows-585dcd7-7db1c333330-4019238418-Release/runtimes/browser-wasm\native -emit-llvm   -l idbfs.js   -IC:/Users/User/AppData/Local/Temp/dotnet-runtime-wasm-windows-585dcd7-7db1c333330-4019238418-Release/runtimes/browser-wasm/native/include/mono-2.0 -Oz -c -o corebindings.o corebindings.c"
2>    [00:00:17.3500222] emcc : The term 'emcc' is not recognized as the name of a cmdlet, function, script file, or operable program. Check
2>    [00:00:17.3500237] the spelling of the name, or if a path was included, verify that the path is correct and try again.
2>    [00:00:17.3500261] At line:1 char:20
2>    [00:00:17.3500274] + $env:PYTHONUTF8=1; emcc  -DENABLE_METADATA_UPDATE=1 -s USE_ZLIB=1 -s  ...
2>    [00:00:17.3500297] +                    ~~~~
2>    [00:00:17.3500312]     + CategoryInfo          : ObjectNotFound: (emcc:String) [], CommandNotFoundException
2>    [00:00:17.3500326]     + FullyQualifiedErrorId : CommandNotFoundException
2>    [00:00:17.3500339]
2>    [00:00:17.3504274] [230/355] cmd /c set "MONO_PATH=" &&set "MONO_PATH=" && C:/Users/User/AppData/Local/Temp/dotnet-runtime-wasm-windows-585dcd7-7db1c333330-4019238418-Release/runtimes/browser-wasm/native/cross/browser-wasm/mono-aot-cross.exe --print-icall-table > icall-table.json`

I have installed the latest version of python from Windows App Store prior to this.

If I run `python --version`  in Power Shell it returns `Python 3.10.2`

From the above errors this line seems interesting:

`The term 'emcc' is not recognized as the name of a cmdlet, function, script file, or operable program.`

**Reply by jeromelaban** [reply]
This is likely caused by a missing component that the bootstrapper did not detect. Try deleting `%USERPROFILE%\.uno` and try building again. The first binlog may give more information.

**Reply by asadullah-refat-89** [reply]
Deleting the `%USERPROFILE%\.uno` folder did the trick. I was able to generate an `aot.profile`. Now pushing to GitHub pages and see if it works. Thanks @jeromelaban

**Reply by asadullah-refat-89** [reply]
> This is likely caused by a missing component that the bootstrapper did not detect. Try deleting `%USERPROFILE%\.uno` and try building again. The first binlog may give more information.

Hey, @jeromelaban This solved my problem. Just checked the build pipelines and it is working like a charm. Thanks for the quick help. Much appreciated.

**Comment by premiaware** [comment]
I have similar issue using last version on Uno Platform 

 error : System.Exception: Failed to run emscripten

I have try to set Build Mode = Diagnostic but no additional info is showed.
I have try to delete %USERPROFILE%\.uno without success

the error is:
2>C:\Users\premiaware\.nuget\packages\uno.wasm.bootstrap\8.0.15\build\Uno.Wasm.Bootstrap.targets(213,3): error : System.Exception: Failed to run emscripten. More details are available in diagnostics logging mode or using the MSBuild /bl switch, see https://aka.platform.uno/msbuild-troubleshoot.
2>C:\Users\premiaware\.nuget\packages\uno.wasm.bootstrap\8.0.15\build\Uno.Wasm.Bootstrap.targets(213,3): error :    at Uno.Wasm.Bootstrap.ShellTask_v1f66b9b15b66efa87ca84d70b6879ddd12d18414.RunPackager() in /_/src/Uno.Wasm.Bootstrap/ShellTask.cs:line 768
2>C:\Users\premiaware\.nuget\packages\uno.wasm.bootstrap\8.0.15\build\Uno.Wasm.Bootstrap.targets(213,3): error :    at Uno.Wasm.Bootstrap.ShellTask_v1f66b9b15b66efa87ca84d70b6879ddd12d18414.Execute() in /_/src/Uno.Wasm.Bootstrap/ShellTask.cs:line 271

You can help me?

**Reply by jeromelaban** [reply]
You can see further errors by going through the link in the doc and the msbuild binlog viewer.

**Reply by premiaware** [reply]
I don't know if I did what you asked me above correctly. now I have this screen you see below, but I don't understand where is the issue.

![image](https://github.com/user-attachments/assets/981621e1-ea71-49f6-9bc9-3e36bd06e9d6)

**Reply by jeromelaban** [reply]
You need to use the the [msbuild binlog viewer](http://msbuildlog.com/) to troubleshoot this further, not the integrated viewer as it does not give enough information.

**Reply by jeromelaban** [reply]
> msbuild ReportAgente.sln /bl

There no error with this particular build, you won't find anything. You need to take a look at the binlog generated by visual studio, if it is the one that is failing. You can post the binlog here as well.

**Reply by premiaware** [reply]
i can see this in logs
<img width="1355" alt="image" src="https://github.com/user-attachments/assets/6529777e-09ec-4e29-a1af-73ddb7704d1a">

I have try to add this in my project

<Project Sdk="Uno.Sdk">
  <PropertyGroup>
    <WasmInitialHeapSize>16777216</WasmInitialHeapSize>
    <EmccInitialHeapSize>32505856</EmccInitialHeapSize>
    

but no fix the issue.
You have other solution?

**Reply by jeromelaban** [reply]
You can take a look at the docs here: https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/runtime-execution-modes.html#increasing-the-initial-memory-size

**Reply by premiaware** [reply]
Now build with success !!!
I don't know how I would have done it without your help, many thanks

One question, in this page 
https://platform.uno/blog/build-net-aot-for-webassembly-in-visual-studio-with-uno-platform/

if i add this:
ItemGroup
	WasmShellExtraEmccFlags Include="-lidbfs.js" 
ItemGroup

the error come back, is possible to build without this ItemGroup ?

Second question:
In the Desktop PC all working and ItemsRepeater is faster, 
In the mobile is very slow and ItemsRepeater when scroll not repaint the page (is not possible to use).

My WebApp is publish as Azure Static Web App.

according to your point of view how come it is so slow on microsoft edge mobile?

**Reply by jeromelaban** [reply]
Great news! This page is very outdated, make sure to use the docs instead.

**Reply by premiaware** [reply]
this is my beta UNO App.

https://gray-rock-03f104903-preview.westeurope.5.azurestaticapps.net/

according to your point of view how come it is so slow on microsoft edge mobile?
in the Desktop pc is ok.

**Reply by jeromelaban** [reply]
In general, Wasm on mobile is not particularly fast, particularly on iOS. Still, you can profile the app with [this configuration](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-profiling.html#cpu-profiling), and the browser tab's performance view.

**Reply by premiaware** [reply]
I am trying to improve and optimise the app.
I have noticed that the option:

<WasmShellEnableThreads>true</WasmShellEnableThreads>.

It works on my local windows pc, but it doesn't work when I send my app to Git which publishes it to Azure Static Web App. 
The app doesn't start it gets stuck in the page with the logo.
could you tell me what could be the cause?

**Reply by jeromelaban** [reply]
We can't tell without any logs, though in general, you can troubleshoot by taking a look at the browser debugger window.

**Reply by premiaware** [reply]
the issue is this:
uno-bootstrap.js:251  Uncaught (in promise) .NET runtime initialization failed (Error: Assert failed: SharedArrayBuffer is not enabled on this page. Please use a modern browser and set Cross-Origin-Opener-Policy and Cross-Origin-Embedder-Policy http headers. See also https://aka.ms/dotnet-wasm-features)


i have try ti add this in web.config

<system.webServer>

```
  <httpProtocol>
    <customHeaders>
      <add name="Cross-Origin-Embedder-Policy" value="require-corp"/>
      <add name="Cross-Origin-Opener-Policy" value="same-origin"/>
      <add name="Cross-Origin-Resource-Policy" value="cross-origin"/>
    </customHeaders>
  </httpProtocol>

```

and this in staticwebapp.config.json
```
{
  "globalHeaders": {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET, POST, PUT, OPTIONS",
    "Cross-Origin-Embedder-Policy": "require-corp",
    "Cross-Origin-Opener-Policy": "same-origin",
    "Cross-Origin-Resource-Policy": "cross-origin"
  },
  "navigationFallback": {

```
My app is hosted in azure static Web App, 
You have idea for fix this issue?

**Reply by premiaware** [reply]
Does anyone have any ideas how to solve this problem?

**Reply by jeromelaban** [reply]
Here's an example of app that is deploying using threading: https://github.com/unoplatform/Uno.RayTraceBenchmark/blob/master/CSharp/Uno/RayTraceBenchmark/RayTraceBenchmark/Platforms/WebAssembly/wwwroot/staticwebapp.config.json. Can you compare with your configuration?

**Reply by premiaware** [reply]
Thank's for support i have try to change my staticwebapp.config.json

[staticwebapp.config.json](https://github.com/user-attachments/files/16872984/staticwebapp.config.json)


but the issue is the same, this is the web page
https://[reportagente.richiesteimmobili.it/?p=b3f7844f-5e85-48b6-bef3-0815b943387a.1](https://reportagente.richiesteimmobili.it/?p=b3f7844f-5e85-48b6-bef3-0815b943387a.1)

**Reply by jeromelaban** [reply]
This configuration is from the documentation, if it does not work for you, there may be something different that is not documented on our side, but we do not know what that could be. 

Make sure to review Microsoft's documentation on Azure Static Apps to determine why the headers are not set.

**Comment by premiaware** [comment]
if i start this command
msbuild ReportAgente.sln /bl

the build terminate with this:
<img width="500" alt="image" src="https://github.com/user-attachments/assets/718ee149-a185-40a7-825e-204c074f05e2">

Where i can find the error in the viewer?
<img width="1376" alt="image" src="https://github.com/user-attachments/assets/a3759c78-00ed-4043-a438-dc99ec69600c">

**Comment by premiaware** [comment]
Is very complex to solve this issue:
i have find this information, but i don't Know is it is usefull

![image](https://github.com/user-attachments/assets/5eece05c-6aa4-47a8-9636-420baaadcfc1)
